### PR TITLE
Add complete codes for an article about deep copy

### PR DESCRIPTION
**Pull request by xenteros**

### CODE REVIEW

File: core-java/src/main/java/com/baeldung/deepcopy/Address.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package com.baeldung.deepcopy;
+
+import java.io.Serializable;
+
+class Address implements Serializable, Cloneable {
```
Comment (KevinGilmore) [inline core-java/src/main/java/com/baeldung/deepcopy/Address.java]: make public

File: core-java/src/main/java/com/baeldung/deepcopy/User.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package com.baeldung.deepcopy;
+
+import java.io.Serializable;
+
+class User implements Serializable, Cloneable {
```
Comment (KevinGilmore) [inline core-java/src/main/java/com/baeldung/deepcopy/User.java]: make public

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+import java.io.IOException;
+
+import org.apache.commons.lang.SerializationUtils;
+import org.junit.Test;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.gson.Gson;
+
+public class DeepCopyTest {
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: rename to DeepCopyUnitTest

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    @Test
+    public void whenCreatingDeepCopyWithCopyConstructor_thenObjectsShouldNotBeSame() {
+
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        User deepCopy = new User(pm);
+
+        assertNotSame(deepCopy, pm);
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: Use the AssertJ assertion:   assertThat(deepCopy).isNotSameAs(pm)

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Test
+    public void whenModifyingOriginalObject_ThenConstructorCopyShouldNotChange() {
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: lowercase "t" in "_then..."

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        address.setCountry("Great Britain");
+
+        assertThat(deepCopy.getAddress().getCountry())
+                .isNotEqualTo(pm.getAddress().getCountry());
+    }
+
+
+    @Test
+    public void shouldDisplayHowLongEachMethodTake() throws CloneNotSupportedException, IOException {
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: I know it's not really a BDD-style test, but let's still come up with a when_then_ name, perhaps whenMakingCopies_thenShowHowLongEachMethodTakes

File: core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package com.baeldung.deepcopy;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.Test;
+
+public class ShallowCopyTest {
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java]: ShallowCopyUnitTest

File: core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+
+    @Test
+    public void whenShallowCoping_thenObjectsShouldNotBeSame() {
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java]: spelling: whenShallowCopying... (missing the "y" in "Copying")

File: core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        User shallowCopy = new User(pm.getFirstName(), pm.getLastName(), pm.getAddress());
+
+        assertThat(shallowCopy)
+                .isNotSameAs(pm);
+    }
+
+    @Test
+    public void whenModifyingOriginalObject_ThenCopyShouldChange() {
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java]: test fails

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+
+        long start = System.currentTimeMillis();
+        for (int i = 0; i < 1000000; i++) {
```
Comment (KevinGilmore) [inline core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: Two notes: First, wouldn't it be good to just use a constant here? And second, one million is kind of a lot for it to be in a unit test that gets run automatically during the build, so let's add @Ignore to this test. If someone wants to run it manually, they can just remove the annotation.

File: core-java/src/main/java/com/baeldung/deepcopy/Address.java (hunk 1 above)
Comment (xenteros) [reply core-java/src/main/java/com/baeldung/deepcopy/Address.java]: Applied

File: core-java/src/main/java/com/baeldung/deepcopy/User.java (hunk 2 above)
Comment (xenteros) [reply core-java/src/main/java/com/baeldung/deepcopy/User.java]: Applied

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java (hunk 3 above)
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: Applied

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java (hunk 4 above)
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: Applied

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java (hunk 5 above)
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: Applied

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java (hunk 6 above)
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: applied

File: core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java (hunk 7 above)
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java]: Applied

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+import com.google.gson.Gson;
+
+public class DeepCopyTest {
+
+    @Test
+    public void whenCreatingDeepCopyWithCopyConstructor_thenObjectsShouldNotBeSame() {
+
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        User deepCopy = new User(pm);
```
[the remark this answers is not in the document]
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: Applied

File: core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java (hunk 9 above)
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/ShallowCopyTest.java]: Fixed

File: core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java (hunk 10 above)
Comment (xenteros) [reply core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java]: Applied

**Pull request merged**